### ISSUE TITLE
Add: Spectral Transfer using Information Geometry

### ISSUE THREAD

**Issue by aj-ptw** [alexandrebarachant/pyRiemann#46]
Please add STIG

http://journal.frontiersin.org/article/10.3389/fnins.2016.00430/full

**Comment by alexandrebarachant** [comment]
Thanks AJ, 
Once you are done, open a pull request.

I can see two usecase :
- Fit a bunch of different pipeline on the same data, and combine preds with STIG. This is typically like the votting classifier from sklearn (your code is based on that)
- Fit different estimators on different subjects, and combine preds with STIG, like in the paper.

so IMO, your code should implement a fit function, but should also be able to take a list of already fitted estimators.

**Comment by aj-ptw** [comment]
I can't understand the first use case, STIG specifically does not train on the same data and the paper surmises that the classifiers must be of the same type. In other words I don't understand why there would be a fit function. I'm thinking of implementing this as just calling predict over and over and never worry about fit.

**Comment by alexandrebarachant** [comment]
Is it a theoritical limitation i.e. the method will not work if the estimators are fitted to the same data or if they are from different type ?

I'm thinking about using this method for ensembling (i.e better than majority voting), like it is done in the original paper: https://arxiv.org/pdf/1303.3257v3.pdf

@nwayt001 any comments ?

**Comment by aj-ptw** [comment]
@alexandrebarachant (or @nwayt001) should i do [follow this advice](https://github.com/nwayt001/Transfer_Learning_Project/blob/master/Online%20Experiment/BCI2000/prog/matlab_RSVP/tools/applySML.m#L4) to:

>These lines remove classifiers who make only positive predictions or only negative predictions; it causes issues with the eigendecomposition if they are not removed; we assign a weighting of zero to these classifiers

**Comment by alexandrebarachant** [comment]
This is a problem when working with prediction instead of classficiation score (like proba). A biased classifier will return always the same class.
This will in return decrease the rank of the prediction matrix and will make the eigendecomposition crash.

you should either throw an error, or remove the line.

**Comment by aj-ptw** [comment]
> you should either throw an error, or remove the line.

@alexandrebarachant i removed the classifiers from the covariance matrix and did essentially [check it out](https://github.com/aj-ptw/pyRiemann/blob/3f183887568cb141385377a52799276895e53cb5/pyriemann/ensemble.py#L243-L268)

At this point i have written all the code that Nick did in his Transfer Learning Project, unfortunately there was no code for the EM step 👎 so going to have to slice and dice what we have to add that.

**Comment by alexandrebarachant** [comment]
that's not related to covariance based classifier. Any classifier can return prediction with probas > 0.5 for all trials.

I think the code can be simplified. if you don't use probas for anithing else than thresholding them to 0.5 then you can directly call the predict function

**Comment by aj-ptw** [comment]
> that's not related to covariance based classifier.

What is _that_ here?

> if you don't use probas for anithing else than thresholding them to 0.5 then you can directly call the predict function

I do that because i have to find the classifier with the largest distance from 0.5 [here](https://github.com/aj-ptw/pyRiemann/blob/3f183887568cb141385377a52799276895e53cb5/pyriemann/ensemble.py#L300-L308) 

> I think the code can be simplified. 

I'm sure it can (and will be) just want to get it working first, then optimize it

**Comment by alexandrebarachant** [comment]
> What is that here?

you can ignore my remark, i think i missunderstood your sentence.
you already removed the bad classifiers ?

> I use pred_probas because i have to find the classifier with the largest distance from 0.5 here

i see. 

keep up the good work :)

**Comment by nwayt001** [comment]
I'll try to address all of these in one go. 

> 4. Are two classifiers training on two separate sessions from the same user considered independent?

> What if I have two classifiers for a user and 10 from other users? Can I use both for a total of 12? Should I pool together the covs and labels to form a cross session classifiers for the user and then only have 11 classifiers in the STIG?

> "...STIG specifically does not train on the same data and the paper surmises that the classifiers must be of the same type...."

> Is it a theoritical limitation i.e. the method will not work if the estimators are fitted to the same data or if they are from different type ?

STIG, which is based on the spectral-meta learning approach from Parisi et al 2014, requires that the classifiers in the ensemble make independent errors. 

It's a common assumption with EEG that data from different subjects trained using the same classifier will produce independent errors. Similarly, different classifiers trained on the same data will produce independent errors (for the most part). So using either classifiers trained on different data or using different classifiers on the same data (or some combination of both) to build you're ensemble should work fine for STIG.

**Comment by aj-ptw** [comment]
@alexandrebarachant should this issue and new feature be called `SML` instead of `STIG`?

**Comment by aj-ptw** [comment]
Almost got this integrated into my Thinker application. I'm thinking that calling predict with an X should return a result right away. Meaning it should be multiplied by a weight in as few calculations as possible. Further I think it best to call `fit` when the user wants to kick off a new fit. We can also do it over a sliding window where every `window` a new _Expectation Maximization_ effort is dispatched (although @nwayt001 I cannot get this working right) however the `v` principal eigenvalue decomposition gives a great weight matrix it seems. Really happy I put a bunch of tests in, caught some bugs and will keep adding tests. I had fun making tests for this, if you see in the source code, Kun and I had to figure out how to generate `n` classifiers fit on different datasets but still be able to send a new dataset through at unit test time that classifiers would respond to with varying degrees of accuracy. They are all above chance, you can check it out! All in all I have really upped my `numpy` ninja skills and am happy with the results of this ensemble classifier so far. I've got 5 classifiers in the cloud ready for the stig, racing to finish the back end for implementing stig in thinker. Thanks for all the help guys!
